File: scripts/src/update-pkhex.ts
#!/bin/node
import fs from 'node:fs';
import AdmZip from 'adm-zip';
import path from 'node:path';
import { spawnSync } from 'child_process';

const releaseUrl = 'https://api.github.com/repos/kwsch/PKHeX/releases/latest';

const pkhexVersionFilepath = '../PKVault.Backend/PKHeX.version';
const pkhexOutputPath = '../PKVault.Backend';

const pkhexRepoFolder = '../tmp-pkhex';

type Release = {
    id: number;
    tag_name: string; // "26.02.27"
    draft: boolean;
    prerelease: boolean;
    tarball_url: string;
    zipball_url: string;
};

/**
 * Update PKHeX from latest release source code.
 */
const act = async () => {
    const currentTag = fs.existsSync(pkhexVersionFilepath)
        ? fs.readFileSync(pkhexVersionFilepath, 'utf8').trim()
        : '';
    console.log('Current PKHeX tag:', currentTag);

    const release = await fetch(releaseUrl).then(res => res.json() as Promise<Release>);

    console.log('Release PKHeX tag:', release.tag_name);

    if (release.tag_name === currentTag) {
        console.log('Already current tag, PKHeX update aborted.');
        return;
    }

    const sourcePath = await fetchAndExtractZipSource(release.zipball_url);

    const pkhexCorePath = path.join(sourcePath, 'PKHeX.Core');

    const commandBase = `dotnet`;
    const commandArgs = `publish ${pkhexCorePath} -o ${pkhexOutputPath}`;
    console.log();
    console.log(commandBase, commandArgs);

    const proc = spawnSync(commandBase, commandArgs.split(' '));

    console.log(`stderr: ${proc.stderr.toString()}`);
    console.log(`stdout: ${proc.stdout.toString()}`);
    console.log();

    fs.writeFileSync(pkhexVersionFilepath, release.tag_name);

    fs.rmSync(pkhexRepoFolder, { recursive: true });
};

const fetchAndExtractZipSource = async (releaseZipUrl: string) => {
    if (fs.existsSync(pkhexRepoFolder)) {
        fs.rmSync(pkhexRepoFolder, { recursive: true });
    }

    console.log('Fetching PKHeX compressed source');

    const zipFileArrayBuffer = await fetch(releaseZipUrl).then(res => res.arrayBuffer());
    const zipFileBuffer = Buffer.from(zipFileArrayBuffer);

    console.log('Unzip PKHeX source');

    const zip = new AdmZip(zipFileBuffer);
    const firstEntry = zip.getEntries()[ 0 ];

    zip.extractAllTo(pkhexRepoFolder);

    return path.join(pkhexRepoFolder, firstEntry.entryName);
};

act();

File: Scripts/update-pkhex/UpdatePKHeX.cs

using System.Diagnostics;
using System.IO.Compression;
using System.Net.Http.Json;
using System.Text;
using System.Text.Json;
using Serilog;

public class UpdatePKHeX
{
    record Release(
        int Id,
        string Tag_name, // "26.02.27"
        bool Draft,
        bool Prerelease,
        string Published_at,
        string Tarball_url,
        string Zipball_url
    );

    const string releaseListUrl = "https://api.github.com/repos/kwsch/PKHeX/releases";
    const string pkhexVersionFilepath = "../PKVault.Core/PKHeX.version";
    const string pkhexOutputPath = "../PKVault.Core";
    const string pkhexRepoFolder = "../tmp-pkhex";

    /**
     * Update PKHeX from latest release source code.
     */
    public static async Task Update()
    {
        var currentTag = File.Exists(pkhexVersionFilepath)
            ? File.ReadAllText(pkhexVersionFilepath).Trim()
            : "";

        Log.Information($"Current PKHeX tag: {currentTag}");

        using var client = new HttpClient();
        client.DefaultRequestHeaders.Add("User-Agent", "PKVault");
        using var res = await client.GetAsync(releaseListUrl);
        using var content = res.Content;
        if ((int)res.StatusCode >= 400)
            throw new Exception($"Error {res.StatusCode} {await content.ReadAsStringAsync()}");

        var releaseList = await content.ReadFromJsonAsync<Release[]>(new JsonSerializerOptions()
        {
            PropertyNameCaseInsensitive = true
        });

        var release = releaseList!
            .Where(r => !r.Draft && !r.Prerelease)
            .OrderByDescending(r => r.Published_at)
            .First();

        Log.Information($"Release PKHeX tag: {release.Tag_name}");

        if (release.Tag_name == currentTag)
        {
            Log.Information($"Already current tag, PKHeX update aborted.");
            return;
        }

        var sourcePath = await FetchAndExtractZipSource(release.Zipball_url);

        var pkhexCorePath = Path.Combine(sourcePath, "PKHeX.Core");

        var commandBase = "dotnet";
        var commandArgs = $"publish {pkhexCorePath} -o {pkhexOutputPath}";
        Console.WriteLine();
        Log.Debug($"{commandBase} {commandArgs}");

        var psi = new ProcessStartInfo
        {
            FileName = commandBase,
            Arguments = commandArgs,
            RedirectStandardError = true,
            RedirectStandardOutput = true,
        };

        using var proc = Process.Start(psi)!;
        await proc.WaitForExitAsync();

        var stderr = await proc.StandardError.ReadToEndAsync();
        var stdout = await proc.StandardOutput.ReadToEndAsync();

        Log.Debug($"stderr: {stderr}");
        Log.Debug($"stdout: {stdout}");
        Console.WriteLine();

        if (proc.ExitCode > 0) {
            Environment.Exit(proc.ExitCode);
        }

        File.WriteAllText(pkhexVersionFilepath, release.Tag_name);

        Directory.Delete(pkhexRepoFolder, true);
    }

    private static async Task<string> FetchAndExtractZipSource(string releaseZipUrl)
    {
        if (Directory.Exists(pkhexRepoFolder))
            Directory.Delete(pkhexRepoFolder, true);

        Log.Information("Fetching PKHeX compressed source");

        using var client = new HttpClient();
        client.DefaultRequestHeaders.Add("User-Agent", "PKVault");
        using var res = await client.GetAsync(releaseZipUrl);
        using var content = res.Content;
        if ((int)res.StatusCode >= 400)
            throw new Exception($"Error {res.StatusCode} {await content.ReadAsStringAsync()}");

        using var stream = content.ReadAsStream();

        Log.Information("Unzip PKHeX source");

        using var zip = await ZipArchive.CreateAsync(stream, ZipArchiveMode.Read, true, Encoding.UTF8);
        var firstEntry = zip.Entries.First();

        await zip.ExtractToDirectoryAsync(pkhexRepoFolder);

        return Path.Combine(pkhexRepoFolder, firstEntry.FullName);
    }
}
